src/sources/apache_metrics/parser.rs

```diff
@@ -0,0 +1,919 @@
+                Err(err) => Some(Err(ParseError {
+                    key: key.to_string(),
+                    err: err.into(),
+                })),
+            },
+            "Scoreboard" => Some(Ok(StatusFieldStatistic::Scoreboard(value))),
+
+            _ => None,
+        }
+    }
+}
+
+/// Parses the text output from Apache's mod_status and returns:
+///
+/// - A list of metrics generated from the output
+/// - A list of parse errors that were encountered
+///
+/// # Arguments
+///
+/// - `payload` - the mod_status output
+/// - `namespace` - the namespace to put the generated metrics in
+/// - `now` - the time the payload was fetched
+/// - `tags` - any base tags to apply to the metrics
+pub fn parse(
+    payload: &str,
+    namespace: &str,
+    now: DateTime<Utc>,
+    tags: Option<&BTreeMap<String, String>>,
+) -> impl Iterator<Item = Result<Metric, ParseError>> {
+    // We use a HashMap rather than a Vector as mod_status has
+    // BusyWorkers/IdleWorkers repeated
+    // https://bz.apache.org/bugzilla/show_bug.cgi?id=63300
+    let parsed = payload
+        .lines()
+        .into_iter()
+        .filter_map(|l| {
+            let mut parts = l.splitn(2, ":");
+            let key = parts.next();
+            let value = parts.next().map(|s| s.trim());
+            match (key, value) {
+                (Some(k), Some(v)) => Some((k, v)),
+                _ => None,
+            }
+        })
+        .collect::<HashMap<_, _>>();
+
+    parsed
+        .iter()
+        .filter_map(|(key, value)| line_to_metrics(key, value, namespace, now, tags))
+        .fold(vec![], |mut acc, v| {
+            match v {
+                Ok(metrics) => metrics.into_iter().for_each(|v| acc.push(Ok(v))),
+                Err(err) => acc.push(Err(err)),
+            };
+            acc
+        })
+        .into_iter()
+}
+
+#[derive(Debug)]
+pub struct ParseError {
+    key: String,
+    err: Box<dyn Error>,
+}
+
+impl fmt::Display for ParseError {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "could not parse value for {}: {}", self.key, self.err)
+    }
+}
+
+impl Error for ParseError {
+    fn source(&self) -> Option<&(dyn Error + 'static)> {
+        Some(self.err.as_ref())
+    }
+}
+
+fn line_to_metrics(
+    key: &str,
+    value: &str,
+    namespace: &str,
+    now: DateTime<Utc>,
+    tags: Option<&BTreeMap<String, String>>,
+) -> Option<Result<Vec<Metric>, ParseError>> {
+    StatusFieldStatistic::from_key_value(key, value).map(|result| {
+        result.map(|statistic| match statistic {
+            StatusFieldStatistic::ServerUptimeSeconds(value) => vec![Metric {
+                name: encode_namespace(namespace, "uptime_seconds_total"),
+                timestamp: Some(now),
+                tags: tags.map(|tags| tags.clone()),
+                kind: MetricKind::Absolute,
+                value: MetricValue::Counter { value },
+            }],
```

Comment:
This looks like there would be some opportunity for refactoring the duplication, but I see there is a few different things going on with `tags` and `value`.